src/pdf_parser/pdf_utils/disambiguate_layout.py

```diff
@@ -0,0 +1,372 @@
+        soft_margin = {
+            "top": unnest_soft_margin,
+            "bottom": unnest_soft_margin,
+            "left": unnest_soft_margin,
+            "right": unnest_soft_margin,
+        }
+        # The loop checks each block for containment within other blocks.
+        # Contained blocks are removed if they have lower confidence scores than their parents;
+        # otherwise, the parent is removed. The process continues until there are no contained blocks.
+        # There are potentially nestings within nestings, hence the complicated loop.
+        # A recursion might be more elegant, leaving it as a TODO.
+        stop_cond = True
+        counter = 0  # num contained blocks in every run through of all pairs to calculate stop
+        # condition.
+        ixs_to_remove = []
+        while stop_cond:
+            for ix, box_1 in enumerate(layout):
+                for ix2, box_2 in enumerate(layout):
+                    if box_1 == box_2:
+                        continue
+                    else:
+                        if box_1.is_in(box_2, soft_margin):
+                            counter += 1
+                            # Remove the box the model is less confident about.
+                            if box_1.score > box_2.score:
+                                remove_ix = ix2
+                            else:
+                                remove_ix = ix
+                            ixs_to_remove.append(remove_ix)
+                # stop condition: no contained blocks
+                if counter == 0:
+                    stop_cond = False
+                counter = 0
+
+        layout_unnested = Layout(
+            [box for index, box in enumerate(layout) if index not in ixs_to_remove]
+        )
+        return layout_unnested
+
+
+# TODO: This is not part of the module and is for a CLI, but placing here for visibility before I edit the CLI.
+def disambiguation_pipeline(
+    image: PpmImageFile,
+    model: Detectron2LayoutModel,
+    restrictive_model_threshold: float,
+    unnest_soft_margin: int,
+    max_overlapping_pixels_vertical: int,
+    max_overlapping_pixels_horizontal: int,
+) -> Layout:
+    """
+    Run the disambiguation pipeline on an image.
+
+    @param image: An image of a PDF page to perform object/block detection on.
+    @param model: Object detection model.
+    @param restrictive_model_threshold: Model confidence to separate blocks into two groups, a "restrictive" group and a "non-restrictive" group.
+    @param unnest_soft_margin: The number of pixels to inflate each box by in each direction when checking for containment for unnesting.
+    @param max_overlapping_pixels_vertical: The maximum number of pixels to allow for vertical overlaps.
+    @param max_overlapping_pixels_horizontal: The maximum number of pixels to allow for horizontal overlaps.
+    @return: A layout object containing only blocks from layoutparser with best effort disambiguation..
+    """
```

Comment:
This function is for a CLI, but put up for visibility for now (am aware of docstring format inconsistency atm :)).